lib/function/arithmetic/log1p.js

```diff
@@ -0,0 +1,105 @@
+'use strict';
+
+var deepMap = require('../../utils/collection/deepMap');
+
+function factory (type, config, load, typed) {
+  var divideScalar = load(require('./divideScalar'));
+  var log = load(require('./log'));
+
+  /**
+   * Calculate the logarithm of a `value+1`.
+   *
+   * For matrices, the function is evaluated element wise.
+   *
+   * Syntax:
+   *
+   *    math.log1p(x)
+   *    math.log1p(x, base)
+   *
+   * Examples:
+   *
+   *    math.log1p(2.5);                 // returns 1.252762968495368
+   *    math.exp(math.log1p(1.4));       // returns 2.4
+   *
+   *    math.pow(10, 4);                 // returns 10000
+   *    math.log1p(9999, 10);            // returns 4
+   *    math.log1p(9999) / math.log(10); // returns 4
+   *
+   * See also:
+   *
+   *    exp, log, log2, log10
+   *
+   * @param {number | BigNumber | Complex | Array | Matrix} x
+   *            Value for which to calculate the logarithm of `x+1`.
+   * @param {number | BigNumber | Complex} [base=e]
+   *            Optional base for the logarithm. If not provided, the natural
+   *            logarithm of `x+1` is calculated.
+   * @return {number | BigNumber | Complex | Array | Matrix}
+   *            Returns the logarithm of `x+1`
+   */
+  var log1p = typed('log1p', {
+    'number': _log1pNumber,
+
+    'Complex': _log1pComplex,
+
+    'BigNumber': function (x) {
+      var y = x.plus(x.constructor.ONE);
+      if (!y.isNegative() || config.predictable) {
+        return y.ln();
+      }
+      else {
+        // downgrade to number, return Complex valued result
+        return _log1pComplex(new type.Complex(x.toNumber(), 0));
+      }
+    },
+
+    'Array | Matrix': function (x) {
+      return deepMap(x, log1p);
+    },
+
+    'any, any': function (x, base) {
+      // calculate logarithm for a specified base, log1p(x, base)
+      return divideScalar(log1p(x), log(base));
+    }
+  });
+
+  /**
+   * Calculate the natural logarithm of a `number+1`
+   * @param {number} x
+   * @returns {number | Complex}
+   * @private
+   */
+  function _log1pNumber(x) {
+    if (x >= -1 || config.predictable) {
+      return (Math.log1p) ? Math.log1p(x) : Math.log(x+1);
+    }
+    else {
+      // negative value -> complex value computation
+      return _log1pComplex(new type.Complex(x, 0));
+    }
+  }
+
+  /**
+   * Calculate the natural logarithm of a complex number + 1
+   * @param {Complex} x
+   * @returns {Complex}
+   * @private
+   */
+  function _log1pComplex(x) {
+    var x_re1p = x.re + 1;
+    return new type.Complex (
+        Math.log(Math.sqrt(x_re1p * x_re1p + x.im * x.im)),
+        Math.atan2(x.im, x_re1p)
+    );
+  }
```

Comment:
Should call out to complex.js here. 
Case where `x.re` is approximately zero should be handled.

Reply:
And `log(sqrt(x))` might be better off using `0.5 * log(x)`.

Reply:
@infusion is it worth implementing these three functions in complex.js?

Reply:
Well, the reason for these functions is numerical stability. Implementing them naively doesn't solve the problem but gives only a shorthand for `exp(z)-1` and `log(1+z)` for the sake of completeness. I don't like to add `logN(z)` (for N ∈ {2, 10, ...}) functions to the complex.js corpus. Any user (like math.js) of the lib is free to add shorthands for this simple scalar multiplication with `1/log(N)`. The other two methods are more interesting and could be candidates for complex.js. We should investigate if there are more sophisticated methods than iterating the infinite series for as much as is needed for double precision, since this could be quite inperformant. 

Additionally, since we work on complex numbers, the same numerical instability could happen on the imaginary part as well, which opens another class of problem other than just supplying the typical API methods known from real numbers.

Reply:
A quick good gave:
```
exp(a + i*b) - 1
= exp(a) * (cos(b) + j*sin(b)) - 1
=   expm1(a)*cos(b) + cosm1(b) + j*exp(a)*sin(b)
```
So if both the real and imaginary parts are small then (real) `exmp1` and `cosm1` are needed for accurate complex `expm1`

Reply:
That looks promising, thanks! :) ES2015 got Math.expm1 and for cosm1 we need to derive a robust implementation. Should we move over to a new complex.js issue for that discussion?